Allow quantifiers within loop invariants

ArenBabikian opened the pull request:
This PR adds support for quantified expressions in loop contracts. ~~This PR depends on #5942.~~

We have included 3 regression tests in this PR. However, two of them are marked as `KNOWNBUG` due to a current limitation of quantifiers that does not allow symbolic ranges for the quantified variable when using the SAT back-end.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

ArenBabikian committed:
Add support for quantifiers in loop contracts

SaswatPadhi commented:
Hi @ArenBabikian,

For the `KNOWNBUG` test cases, could you please try the SMT backend? CPROVER SMT2 is not expected to succeed if the SAT backend does not, but we might have some luck with Z3.

- Try running CBMC with `--z3` locally
- For CI, replace `KNOWNBUG` with `CORE smt-backend broken-cprover-smt-backend`

SaswatPadhi committed:
Refactor and cleanup

We do not need to do quantifier replacement within ASSIGNS annotation.
The replacement maps for ENSURES and REQUIRES clauses should also be
maintained independently.

SaswatPadhi commented:
Marking this as draft for now.

The failure traces look like the one from #6106, but they don't use `char*`. Need to take a closer look, and may be #6106 is due to some issue with contracts -> assertion/assumption translation and not `char *` or `__CPROVER_w_ok`.

martin-cs commented on regression/contracts/quantifiers-loop-01/main.c:
```diff
@@ -0,0 +1,20 @@
+      (0 <= i) && (i <= N) &&
+      __CPROVER_forall {
+        int k;
+        (0 <= k && k < N) ==> 1 == 1
```
Is the `1 == 1` so that it could simplify away completely intentional?

feliperodri replied:
Shouldn't the invariant be the following?
```c
      __CPROVER_forall {
         int k;
         (0 <= k && k < N) ==> a[k] = 1
      }
```

SaswatPadhi replied:
`(0 <= k && k < N) ==> a[k] = 1` wouldn't hold, but I am adding `(0 <= k && k < i) ==> a[k] = 1` to address this concern.

martin-cs commented on regression/contracts/quantifiers-loop-01/test.desc:
```diff
@@ -0,0 +1,11 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test case checks the handling of a `forall` quantifier within a loop invariant.
```
If it is, then please can you expand the description here?

SaswatPadhi replied:
Added some description.

martin-cs reviewed (approved):
This doesn't look too controversial...

feliperodri commented on regression/contracts/quantifiers-loop-02/main.c:
```diff
@@ -0,0 +1,32 @@
+  });
+  // clang-format on
+
+  int k;
+  __CPROVER_assume(0 <= k && k < N);
+  assert(a[k] == 1);
```
Why do we need this block here? It seems redundant.

SaswatPadhi replied:
I kept it around just to nondeterministically check the array without invoking quantifiers at all, and ensure consistency between both the quantified interpretation and this non-det interpretation.

feliperodri commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -213,12 +225,10 @@ void code_contractst::add_quantified_variable(
   {
     // When a quantifier expression is found,
-    // 1. get quantified variables
+    // for each quantified variable ...
```
If you decide to remove `1.`, update the following comments too; otherwise, leave it here.

SaswatPadhi replied:
Done; I updated the comments.

feliperodri commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -127,9 +137,11 @@ void code_contractst::check_apply_invariants(
-  code_assumet assumption{invariant};
-  assumption.add_source_location() = loop_head->source_location;
-  converter.goto_convert(assumption, havoc_code, mode);
+  {
+    code_assumet assumption{invariant_expr()};
+    assumption.add_source_location() = loop_head->source_location;
+    converter.goto_convert(assumption, havoc_code, mode);
+  }
```
Why are you adding `{...}`?

SaswatPadhi replied:
For scoping variables.

The invariant gets expanded to multiple assertions and assumptions. I scope these, for example the `assumption` object in this case, so they don't pollute the scope of the overall function. They are immediately compiled to goto instructions and added to `havoc_code`.
